include/linux/bpf.h

```diff
@@ -58,119 +65,172 @@ enum bpf_cmd_id
     BPF_OBJ_GET,
     BPF_PROG_ATTACH,
     BPF_PROG_DETACH,
+    BPF_PROG_TEST_RUN,
     BPF_PROG_GET_NEXT_ID,
     BPF_MAP_GET_NEXT_ID,
-    BPF_LINK_GET_NEXT_ID,
     BPF_PROG_GET_FD_BY_ID,
     BPF_MAP_GET_FD_BY_ID,
-    BPF_MAP_LOOKUP_AND_DELETE_ELEM,
-    BPF_LINK_GET_FD_BY_ID,
     BPF_OBJ_GET_INFO_BY_FD,
-    BPF_LINK_DETACH,
-    BPF_PROG_BIND_MAP,
-    BPF_PROG_TEST_RUN,
+    BPF_MAP_LOOKUP_AND_DELETE_ELEM = 21,
+    BPF_LINK_GET_FD_BY_ID = 30,
+    BPF_LINK_GET_NEXT_ID,
+    BPF_LINK_DETACH = 34,
+    BPF_PROG_BIND_MAP = 35,
     BPF_PROG_RUN = BPF_PROG_TEST_RUN,
 };
 
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(default : 4820) // reject implicit padding
+#endif
+
+/// Attributes used by BPF_MAP_CREATE.
+typedef struct
+{
+    enum bpf_map_type map_type; ///< Type of map to create.
+    uint32_t key_size;          ///< Size in bytes of keys.
+    uint32_t value_size;        ///< Size in bytes of values.
+    uint32_t max_entries;       ///< Maximum number of entries in the map.
+    uint32_t map_flags;         ///< Flags (currently 0).
+} sys_bpf_map_create_attr_t;
+
+typedef struct
+{
+    uint32_t map_fd; ///< File descriptor of map.
+    uint32_t _pad0;
+    uint64_t key;   ///< Pointer to key to look up.
+    uint64_t value; ///< Pointer to value.
+    uint64_t flags; ///< Not supported.
+} sys_bpf_map_lookup_attr_t;
+
+typedef struct
+{
+    uint32_t map_fd; ///< File descriptor of map.
+    uint32_t _pad0;
+    uint64_t key; ///< Pointer to key to delete.
+} sys_bpf_map_delete_attr_t;
+
+typedef struct
+{
+    uint32_t map_fd; ///< File descriptor of map.
+    uint32_t _pad0;
+    uint64_t key;      ///< Pointer to key to look up.
+    uint64_t next_key; ///< Pointer to next key.
+} sys_bpf_map_next_key_attr_t;
+
+typedef struct
+{
+    uint32_t start_id; ///< ID to look for an ID after. The start_id need not exist.
+    uint32_t next_id;  ///< On return, contains the next ID.
+} sys_bpf_map_next_id_attr_t;
+
+typedef struct
+{
+    enum bpf_prog_type prog_type; ///< Program type to use for loading the program.
+    uint32_t insn_cnt;            ///< Number of instructions in the array.
+    uint64_t insns;               ///< Array of instructions
+    uint64_t license;      ///< Optional pointer to a string specifying the license (currently ignored on Windows).
+    uint32_t log_level;    ///< Logging level (currently ignored on Windows).
+    uint32_t log_size;     ///< Size in bytes of the log buffer.
+    uint64_t log_buf;      ///< Pointer to a buffer in which log info can be written.
+    uint32_t kern_version; ///< Kernel version (currently ignored on Windows).
+    uint32_t prog_flags;   ///< Not supported.
```

Comment:
How does "not supported" differ from "currently ignored on Windows"?

Reply:
Not supported means that the wrapper will return an error if the value is non-zero. Currently ignored means it'll accept any value. The latter is correct for license and kern_version I think, but I'll probably send a change to reject certain values of log_level. Otherwise it's very hard to probe for extended log support from the library side.

Reply:
I tried to clarify this by saying "Not supported, must be zero." instead.

Reply:
Ok